### PR TITLE
Clarify `endIndex` behavior of `partitioningIndex(where:)` in documentation (Resolves #170)

### LINKED ISSUE #170: Clarify `endIndex` behavior of `partitioningIndex(where:)` in documentation
https://github.com/apple/swift-algorithms/blob/e2fa131d62604027889fcf87bb4d024e6576ac66/Sources/Algorithms/Partition.swift#L171-L182

In the documentation for `partitioningIndex(where:)`, it only mentions what happens when the predicate returns `true`. But, if the `predicate` returns `false` in all cases, it is unclear what the return value is. The documentation should be written so that it is clear that the return value is `endIndex` if `belongsInSecondPartition` returns `false` for all elements.

I think this should at least be reflected in the “Returns” section of the inline documentation, but could probably be made more clear by rephrasing the first few sentences. The first sentence makes it sound like this is identical to `firstIndex(where:)`, but the second and third sentences make the behavior more clear, indicating that this function is only to be used on collections that have already been partitioned.

### PR DESCRIPTION
**Pull request by mdznr**
Resolves https://github.com/apple/swift-algorithms/issues/170

### Checklist
- [x] I've added at least one test that validates that my change is working, if appropriate
- [x] I've followed the code style of the rest of the project
- [x] I've read the [Contribution Guidelines](../blob/main/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary

### CODE REVIEW

File: Sources/Algorithms/Partition.swift [hunk 1]
```diff
@@ -168,18 +168,18 @@ extension MutableCollection where Self: BidirectionalCollection {
 //===----------------------------------------------------------------------===//
 
 extension Collection {
-  /// Returns the index of the first element in the collection that matches
-  /// the predicate.
+  /// Returns the start index of the second partition of a collection already
+  /// partitioned according to the predicate.
```
Comment (natecook1000) [inline Sources/Algorithms/Partition.swift:172]: I agree with @xwu that we don't want to lose the more tersely stated precondition below. What would you think about the following edit to remove the redundancy?

```suggestion
  /// Returns the start index of the partition of a collection that matches 
  /// the given predicate.
```

File: Sources/Algorithms/Partition.swift [hunk 2]
```diff
@@ -168,18 +168,18 @@ extension MutableCollection where Self: BidirectionalCollection {
...
   ///
-  /// The collection must already be partitioned according to the predicate.
```
[the remark this answers is not in the document]
Comment (natecook1000) [reply Sources/Algorithms/Partition.swift:174]: Can you add that sentence back, here? Reading it again I think we might also want a counter statement, that calling `partitioningIndex` on a collection that isn't ordered correctly may produce an incorrect result.

File: Sources/Algorithms/Partition.swift [hunk 3]
```diff
@@ -168,18 +168,18 @@ extension MutableCollection where Self: BidirectionalCollection {
...
   /// That is, there should be an index `i` where for every element in
   /// `collection[..<i]` the predicate is `false`, and for every element in
   /// `collection[i...]` the predicate is `true`.
   ///
   /// - Parameter belongsInSecondPartition: A predicate that partitions the
   ///   collection.
   /// - Returns: The index of the first element in the collection for which
-  ///   `predicate` returns `true`.
+  ///   `predicate` returns `true`, or `endIndex` if `predicate` returns `false`
+  ///   for all elements.
```
[the remark this answers is not in the document]
Comment (natecook1000) [reply Sources/Algorithms/Partition.swift:182]: How about this, which covers empty collections as well:
```suggestion
  ///   `predicate` returns `true`, or `endIndex` if there are no elements
  ///   for which `predicate` returns `true`.
```

In addition, empty collections only have one index, so I'm not sure how much confusion there can be.

File: Tests/SwiftAlgorithmsTests/PartitionTests.swift [hunk 4]
```diff
@@ -92,6 +92,26 @@ final class PartitionTests: XCTestCase {
...
+  
+  func testPartitioningIndexWithOneEmptyPartition() {
+    let input: Range<Int> = (0 ..< 10)
+    
+    let a = input.partitioningIndex(where: { $0 > 10 })
+    XCTAssertEqual(a, input.endIndex)
+    
+    let b = input.partitioningIndex(where: { $0 >= 0 })
+    XCTAssertEqual(b, input.startIndex)
+  }
```
Comment (natecook1000) [inline Tests/SwiftAlgorithmsTests/PartitionTests.swift:113]: Thanks for these tests! 👏🏻

**Review by natecook1000 — APPROVED** [review]

**Comment by natecook1000** [comment]
@swift-ci Please test

**Comment by natecook1000** [comment]
@swift-ci Please test

**Review by valerio-123 — COMMENTED** [review]
Mamma

**Pull request merged**